File: contracts/wasm/fairauction/go/fairauction/fairauction.go
// Copyright 2020 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

package fairauction

import (
	"github.com/iotaledger/wasp/packages/wasmvm/wasmlib/go/wasmlib"
	"github.com/iotaledger/wasp/packages/wasmvm/wasmlib/go/wasmlib/wasmtypes"
)

const (
	// default duration is 60 min
	DurationDefault = 60
	// minimum duration is 1 min
	DurationMin = 1
	// maximum duration is 120 min
	DurationMax          = 120
	MaxDescriptionLength = 150
	OwnerMarginDefault   = 50
	OwnerMarginMin       = 5
	OwnerMarginMax       = 100
	DustDeposit          = 1
)

func funcStartAuction(ctx wasmlib.ScFuncContext, f *StartAuctionContext) {
	auctionNFT := f.Params.Nft().Value()
	nfts := ctx.Allowance().NftIDs()
	for _, nft := range nfts {
		if nft.String() == auctionNFT.String() {
			goto exist
		}
	}
	ctx.Panic("Missing auction tokens")

exist:
	// Any transfer to the Smart Contract is implemented as a transfer of assets to the caller's L2 account,
	// with an allowance to those assets passed to the Smart Contract, so Smart Contract needs to actively
	// transfer the for sale NFT to Smart Contract account
	transfer := wasmlib.NewScTransferIotas(DustDeposit)
	transfer.AddNFT(&auctionNFT)
	ctx.TransferAllowed(ctx.AccountID(), transfer, false)

	minimumBid := f.Params.MinimumBid().Value()

	// duration in minutes
	duration := f.Params.Duration().Value()
	if duration == 0 {
		duration = DurationDefault
	}
	if duration < DurationMin {
		duration = DurationMin
	}
	if duration > DurationMax {
		duration = DurationMax
	}

	description := f.Params.Description().Value()
	if description == "" {
		description = "N/A"
	}
	if len(description) > MaxDescriptionLength {
		ss := description[:MaxDescriptionLength]
		description = ss + "[...]"
	}

	ownerMargin := f.State.OwnerMargin().Value()
	if ownerMargin == 0 {
		ownerMargin = OwnerMarginDefault
	}

	// need at least 1 iota (dust deposit) to run SC
	margin := minimumBid * ownerMargin / 1000
	if margin == 0 {
		margin = DustDeposit
	}
	deposit := ctx.Allowance().Iotas()
	if deposit < margin {
		ctx.Panic("Insufficient deposit")
	}

	currentAuction := f.State.Auctions().GetAuction(auctionNFT)
	if currentAuction.Exists() {
		ctx.Panic("Auction for this token already exists")
	}

	auction := &Auction{
		Creator:       ctx.Caller(),
		Deposit:       deposit,
		Description:   description,
		Duration:      duration,
		HighestBid:    0,
		HighestBidder: ctx.Caller(),
		MinimumBid:    minimumBid,
		OwnerMargin:   ownerMargin,
		Nft:           auctionNFT,
		WhenStarted:   ctx.Timestamp(),
	}
	currentAuction.SetValue(auction)

	fa := ScFuncs.FinalizeAuction(ctx)
	fa.Params.Nft().SetValue(auction.Nft)
	fa.Func.Delay(duration * 60).Post()
}

func funcFinalizeAuction(ctx wasmlib.ScFuncContext, f *FinalizeAuctionContext) {
	auctionNFT := f.Params.Nft().Value()
	currentAuction := f.State.Auctions().GetAuction(auctionNFT)
	ctx.Require(currentAuction.Exists(), "Missing auction info")
	auction := currentAuction.Value()
	if auction.HighestBid == 0 {
		ctx.Log("No one bid on " + auctionNFT.String())
		ownerFee := auction.MinimumBid * auction.OwnerMargin / 1000
		if ownerFee == 0 {
			ownerFee = 1
		}
		// finalizeAuction request token was probably not confirmed yet
		transferIotas(ctx, ctx.ContractCreator(), ownerFee-1)
		transferNFT(ctx, auction.Creator, auction.Nft)
		transferIotas(ctx, auction.Creator, auction.Deposit-ownerFee)
		return
	}

	ownerFee := auction.HighestBid * auction.OwnerMargin / 1000
	if ownerFee == 0 {
		ownerFee = 1
	}

	// return staked bids to losers
	bids := f.State.Bids().GetBids(auctionNFT)
	bidderList := f.State.BidderList().GetBidderList(auctionNFT)
	size := bidderList.Length()
	for i := uint32(0); i < size; i++ {
		loser := bidderList.GetAgentID(i).Value()
		if loser != auction.HighestBidder {
			bid := bids.GetBid(loser).Value()
			transferIotas(ctx, loser, bid.Amount)
		}
	}

	// finalizeAuction request token was probably not confirmed yet
	transferIotas(ctx, ctx.ContractCreator(), ownerFee-1)
	transferNFT(ctx, auction.HighestBidder, auction.Nft)
	transferIotas(ctx, auction.Creator, auction.Deposit+auction.HighestBid-ownerFee)
}

func funcPlaceBid(ctx wasmlib.ScFuncContext, f *PlaceBidContext) {
	bidAmount := ctx.Allowance().Iotas()
	ctx.Require(bidAmount > 0, "Missing bid amount")

	token := f.Params.Nft().Value()
	currentAuction := f.State.Auctions().GetAuction(token)
	ctx.Require(currentAuction.Exists(), "Missing auction info")

	auction := currentAuction.Value()
	bids := f.State.Bids().GetBids(token)
	bidderList := f.State.BidderList().GetBidderList(token)
	caller := ctx.Caller()
	currentBid := bids.GetBid(caller)
	if currentBid.Exists() {
		ctx.Log("Upped bid from: " + caller.String())
		bid := currentBid.Value()
		bidAmount += bid.Amount
		bid.Amount = bidAmount
		bid.Timestamp = ctx.Timestamp()
		currentBid.SetValue(bid)
	} else {
		ctx.Require(bidAmount >= auction.MinimumBid, "Insufficient bid amount")
		ctx.Log("New bid from: " + caller.String())
		index := bidderList.Length()
		bidderList.AppendAgentID().SetValue(caller)
		bid := &Bid{
			Index:     index,
			Amount:    bidAmount,
			Timestamp: ctx.Timestamp(),
		}
		currentBid.SetValue(bid)
	}
	if bidAmount > auction.HighestBid {
		ctx.Log("New highest bidder")
		auction.HighestBid = bidAmount
		auction.HighestBidder = caller
		currentAuction.SetValue(auction)
	}
}

func funcSetOwnerMargin(ctx wasmlib.ScFuncContext, f *SetOwnerMarginContext) {
	ownerMargin := f.Params.OwnerMargin().Value()
	if ownerMargin < OwnerMarginMin {
		ownerMargin = OwnerMarginMin
	}
	if ownerMargin > OwnerMarginMax {
		ownerMargin = OwnerMarginMax
	}
	f.State.OwnerMargin().SetValue(ownerMargin)
}

func viewGetAuctionInfo(ctx wasmlib.ScViewContext, f *GetAuctionInfoContext) {
	token := f.Params.Nft().Value()
	currentAuction := f.State.Auctions().GetAuction(token)
	if !currentAuction.Exists() {
		ctx.Panic("Missing auction info")
	}

	auction := currentAuction.Value()
	f.Results.Creator().SetValue(auction.Creator)
	f.Results.Deposit().SetValue(auction.Deposit)
	f.Results.Description().SetValue(auction.Description)
	f.Results.Duration().SetValue(auction.Duration)
	f.Results.HighestBid().SetValue(auction.HighestBid)
	f.Results.HighestBidder().SetValue(auction.HighestBidder)
	f.Results.MinimumBid().SetValue(auction.MinimumBid)
	f.Results.OwnerMargin().SetValue(auction.OwnerMargin)
	f.Results.Nft().SetValue(auction.Nft)
	f.Results.WhenStarted().SetValue(auction.WhenStarted)

	bidderList := f.State.BidderList().GetBidderList(token)
	f.Results.Bidders().SetValue(bidderList.Length())
}

func transferIotas(ctx wasmlib.ScFuncContext, agent wasmtypes.ScAgentID, amount uint64) {
	if agent.IsAddress() {
		// send back to original Tangle address
		ctx.Send(agent.Address(), wasmlib.NewScTransferIotas(amount))
		return
	}

	// TODO not an address, deposit into account on chain
	ctx.Send(agent.Address(), wasmlib.NewScTransferIotas(amount))
}

func transferNFT(ctx wasmlib.ScFuncContext, agent wasmtypes.ScAgentID, nft wasmtypes.ScNftID) {
	if agent.IsAddress() {
		// send back to original Tangle address
		ctx.Send(agent.Address(), wasmlib.NewScTransferNFT(&nft))
		return
	}

	// TODO not an address, deposit into account on chain
	ctx.Send(agent.Address(), wasmlib.NewScTransferNFT(&nft))
}

File: contracts/wasm/fairauction/ts/fairauction/fairauction.ts
// Copyright 2020 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

import * as wasmlib from "wasmlib"
import * as wasmtypes from "wasmlib/wasmtypes";
import * as sc from "./index";

const DURATION_DEFAULT: u32 = 60;
const DURATION_MIN: u32 = 1;
const DURATION_MAX: u32 = 120;
const MAX_DESCRIPTION_LENGTH: i32 = 150;
const OWNER_MARGIN_DEFAULT: u64 = 50;
const OWNER_MARGIN_MIN: u64 = 5;
const OWNER_MARGIN_MAX: u64 = 100;

export function funcFinalizeAuction(ctx: wasmlib.ScFuncContext, f: sc.FinalizeAuctionContext): void {
    let token = f.params.token().value();
    let currentAuction = f.state.auctions().getAuction(token);
    ctx.require(currentAuction.exists(), "Missing auction info");
    let auction = currentAuction.value();
    if (auction.highestBid == 0) {
        ctx.log("No one bid on " + token.toString());
        let ownerFee = auction.minimumBid * auction.ownerMargin / 1000;
        if (ownerFee == 0) {
            ownerFee = 1;
        }
        // finalizeAuction request token was probably not confirmed yet
        transferIotas(ctx, ctx.contractCreator(), ownerFee - 1);
        transferTokens(ctx, auction.creator, auction.token, auction.numTokens);
        transferIotas(ctx, auction.creator, auction.deposit - ownerFee);
        return;
    }

    let ownerFee = auction.highestBid * auction.ownerMargin / 1000;
    if (ownerFee == 0) {
        ownerFee = 1;
    }

    // return staked bids to losers
    let bids = f.state.bids().getBids(token);
    let bidderList = f.state.bidderList().getBidderList(token);
    let size = bidderList.length();
    for (let i: u32 = 0; i < size; i++) {
        let loser = bidderList.getAgentID(i).value();
        if (!loser.equals(auction.highestBidder)) {
            let bid = bids.getBid(loser).value();
            transferIotas(ctx, loser, bid.amount);
        }
    }

    // finalizeAuction request token was probably not confirmed yet
    transferIotas(ctx, ctx.contractCreator(), ownerFee - 1);
    transferTokens(ctx, auction.highestBidder, auction.token, auction.numTokens);
    transferIotas(ctx, auction.creator, auction.deposit + auction.highestBid - ownerFee);
}

export function funcPlaceBid(ctx: wasmlib.ScFuncContext, f: sc.PlaceBidContext): void {
    let bidAmount = ctx.allowance().iotas();
    ctx.require(bidAmount > 0, "Missing bid amount");

    let token = f.params.token().value();
    let currentAuction = f.state.auctions().getAuction(token);
    ctx.require(currentAuction.exists(), "Missing auction info");

    let auction = currentAuction.value();
    let bids = f.state.bids().getBids(token);
    let bidderList = f.state.bidderList().getBidderList(token);
    let caller = ctx.caller();
    let currentBid = bids.getBid(caller);
    if (currentBid.exists()) {
        ctx.log("Upped bid from: " + caller.toString());
        let bid = currentBid.value();
        bidAmount += bid.amount;
        bid.amount = bidAmount;
        bid.timestamp = ctx.timestamp();
        currentBid.setValue(bid);
    } else {
        ctx.require(bidAmount >= auction.minimumBid, "Insufficient bid amount");
        ctx.log("New bid from: " + caller.toString());
        let index = bidderList.length();
        bidderList.appendAgentID().setValue(caller);
        let bid = new sc.Bid();
        bid.index = index;
        bid.amount = bidAmount;
        bid.timestamp = ctx.timestamp();
        currentBid.setValue(bid);
    }
    if (bidAmount > auction.highestBid) {
        ctx.log("New highest bidder");
        auction.highestBid = bidAmount;
        auction.highestBidder = caller;
        currentAuction.setValue(auction);
    }
}

export function funcSetOwnerMargin(ctx: wasmlib.ScFuncContext, f: sc.SetOwnerMarginContext): void {
    let ownerMargin = f.params.ownerMargin().value();
    if (ownerMargin < OWNER_MARGIN_MIN) {
        ownerMargin = OWNER_MARGIN_MIN;
    }
    if (ownerMargin > OWNER_MARGIN_MAX) {
        ownerMargin = OWNER_MARGIN_MAX;
    }
    f.state.ownerMargin().setValue(ownerMargin);
}

export function funcStartAuction(ctx: wasmlib.ScFuncContext, f: sc.StartAuctionContext): void {
    let token = f.params.token().value();
    let numTokens = ctx.allowance().balance(token);
    if (numTokens.isZero()) {
        ctx.panic("Missing auction tokens");
    }

    let minimumBid = f.params.minimumBid().value();

    // duration in minutes
    let duration = f.params.duration().value();
    if (duration == 0) {
        duration = DURATION_DEFAULT;
    }
    if (duration < DURATION_MIN) {
        duration = DURATION_MIN;
    }
    if (duration > DURATION_MAX) {
        duration = DURATION_MAX;
    }

    let description = f.params.description().value();
    if (description == "") {
        description = "N/A".toString();
    }
    if (description.length > MAX_DESCRIPTION_LENGTH) {
        description = description.slice(0,MAX_DESCRIPTION_LENGTH) + "[...]";
    }

    let ownerMargin = f.state.ownerMargin().value();
    if (ownerMargin == 0) {
        ownerMargin = OWNER_MARGIN_DEFAULT;
    }

    // need at least 1 iota to run SC
    let margin = minimumBid * ownerMargin / 1000;
    if (margin == 0) {
        margin = 1;
    }
    let deposit = ctx.allowance().iotas();
    if (deposit < margin) {
        ctx.panic("Insufficient deposit");
    }

    let currentAuction = f.state.auctions().getAuction(token);
    if (currentAuction.exists()) {
        ctx.panic("Auction for this token token already exists");
    }

    let auction = new sc.Auction();
    auction.creator = ctx.caller();
    auction.token = token;
    auction.deposit = deposit;
    auction.description = description;
    auction.duration = duration;
    auction.highestBid = 0;
    auction.highestBidder = wasmtypes.agentIDFromBytes([]);
    auction.minimumBid = minimumBid;
    auction.numTokens = numTokens.uint64();
    auction.ownerMargin = ownerMargin;
    auction.whenStarted = ctx.timestamp();
    currentAuction.setValue(auction);

    let fa = sc.ScFuncs.finalizeAuction(ctx);
    fa.params.token().setValue(auction.token);
    fa.func.delay(duration * 60).post();
}

export function viewGetInfo(ctx: wasmlib.ScViewContext, f: sc.GetInfoContext): void {
    let token = f.params.token().value();
    let currentAuction = f.state.auctions().getAuction(token);
    ctx.require(currentAuction.exists(), "Missing auction info");

    let auction = currentAuction.value();
    f.results.token().setValue(auction.token);
    f.results.creator().setValue(auction.creator);
    f.results.deposit().setValue(auction.deposit);
    f.results.description().setValue(auction.description);
    f.results.duration().setValue(auction.duration);
    f.results.highestBid().setValue(auction.highestBid);
    f.results.highestBidder().setValue(auction.highestBidder);
    f.results.minimumBid().setValue(auction.minimumBid);
    f.results.numTokens().setValue(auction.numTokens);
    f.results.ownerMargin().setValue(auction.ownerMargin);
    f.results.whenStarted().setValue(auction.whenStarted);

    let bidderList = f.state.bidderList().getBidderList(token);
    f.results.bidders().setValue(bidderList.length());
}

function transferIotas(ctx: wasmlib.ScFuncContext, agent: wasmlib.ScAgentID, amount: u64): void {
    if (agent.isAddress()) {
        // send back to original Tangle address
        ctx.send(agent.address(), wasmlib.ScTransfer.iotas(amount));
        return;
    }

    // TODO not an address, deposit into account on chain
    ctx.send(agent.address(), wasmlib.ScTransfer.iotas(amount));
}

function transferTokens(ctx: wasmlib.ScFuncContext, agent: wasmlib.ScAgentID, token: wasmlib.ScTokenID, amount: u64): void {
    const bigAmount = wasmtypes.ScBigInt.fromUint64(amount);
    if (agent.isAddress()) {
        // send back to original Tangle address
        ctx.send(agent.address(), wasmlib.ScTransfer.tokens(token, bigAmount));
        return;
    }

    // TODO not an address, deposit into account on chain
    ctx.send(agent.address(), wasmlib.ScTransfer.tokens(token, bigAmount));
}
